### PR TITLE
WT-7392 Added evicted flag to dhandle for use by session sweep.

### PR DESCRIPTION
**Pull request by ddanderson**

### CODE REVIEW

**Comment by ddanderson** [comment]
@sulabhM, here's the approach I was thinking of in the ticket.  I'm not sure that there isn't a better place to set the flag during eviction.

File: src/session/session_dhandle.c [hunk 1]
```diff
@@ -379,15 +378,16 @@ __session_dhandle_sweep(WT_SESSION_IMPL *session)
...
+         * Only discard handles that are dead or dying and, in the case of btrees, have been
+         * evicted. These checks are not done with any locks in place, other than the data handle
+         * reference, so we cannot peer past what is in the dhandle directly.
+         */
         if (dhandle != session->dhandle && dhandle->session_inuse == 0 &&
           (WT_DHANDLE_INACTIVE(dhandle) ||
-            (dhandle->timeofdeath != 0 && now - dhandle->timeofdeath > conn->sweep_idle_time) ||
-            empty_btree)) {
+            (dhandle->timeofdeath != 0 && now - dhandle->timeofdeath > conn->sweep_idle_time)) &&
+          (!WT_DHANDLE_BTREE(dhandle) || F_ISSET(dhandle, WT_DHANDLE_EVICTED))) {
```
Comment (sulabhM) [inline src/session/session_dhandle.c:390]: Earlier we did a session sweep if:
`in_use ==0 AND (handle_inactive OR handle_expired OR empty_btree)`

Now with the change, it looks like:
`in_use ==0 AND (handle_inactive OR handle_expired) AND (not-btree OR handle-marked-evicted)`

Effectively we are saying that for btree handles, the handle needs to be inactive or expired, but also it needs to have been marked evicted.

I don't see an issue with that, it seems correct to me. But not getting it right will prevent handles from being swept.

Let's do a second review too, just to make sure we have got this right.

**Review by sulabhM — APPROVED** [review]
LGTM. Please get a second round of review. I recommend Sue.

**Comment by ddanderson** [comment]
@sulabhM, the logic change to sweep only when a tree is evicted comes from [WT-4458](https://jira.mongodb.org/browse/WT-4458).  The ticket title/description suggests we only sweep when a tree has been evicted, but the change to the code made in that ticket produced incorrect logic IMHO.  I think this restores the intention of that ticket.  When I merge, I'll fix the git comment so this is noted.  Yes, I'll have Sue review, good idea.

**Comment by ddanderson** [comment]
@sueloverso, can you please review this change to how we determine to discard a dhandle?

File: src/session/session_dhandle.c (hunk 1 above)
Comment (sueloverso) [reply src/session/session_dhandle.c:390]: I think a review of the original ticket that added `empty_btree` might help drive whether it should be `and` or `or`.
Thanks for the good comment, as it hopefully makes clear so no one looks in the btree at a later date.

**Review by sueloverso — APPROVED** [review]
LGTM once the 'and/or' comment is resolved.

**Comment by sulabhM** [comment]
> @sulabhM, the logic change to sweep only when a tree is evicted comes from WT-4458. The ticket title/description suggests we only sweep when a tree has been evicted, but the change to the code made in that ticket produced incorrect logic IMHO. I think this restores the intention of that ticket. When I merge, I'll fix the git comment so this is noted. Yes, I'll have Sue review, good idea.

Sounds, good. I remember you investigated the history as part of this ticket and it looked like WT-4458 did not do the right thing. This change should be good to go. Thanks for a deep dive into this one.

**Pull request merged**